ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -2226,30 +2229,305 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
 	return add_attribute(head, PKCS11_CKA_VALUE, data, key_length);
 }
 
+static enum pkcs11_rc set_private_key_data_rsa(struct obj_attrs **head,
+					       void *data,
+					       size_t key_size)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	int mbedtls_rc = 0;
+	uint32_t key_bits = 0;
+	uint32_t size = 0;
+	uint32_t buffer_size = 0;
+	void *buffer = NULL;
+	mbedtls_pk_context pk = { };
+	mbedtls_rsa_context *rsa = NULL;
+	mbedtls_mpi n = { };
+	mbedtls_mpi e = { };
+	mbedtls_mpi d = { };
+	mbedtls_mpi p = { };
+	mbedtls_mpi q = { };
+
+	rc = get_u32_attribute(*head, PKCS11_CKA_MODULUS_BITS, &key_bits);
+	if (rc && rc != PKCS11_RV_NOT_FOUND)
+		return rc;
+
+	if (remove_empty_attribute(head, PKCS11_CKA_MODULUS) ||
+	    remove_empty_attribute(head, PKCS11_CKA_PUBLIC_EXPONENT) ||
+	    remove_empty_attribute(head, PKCS11_CKA_PRIVATE_EXPONENT) ||
+	    remove_empty_attribute(head, PKCS11_CKA_PRIME_1) ||
+	    remove_empty_attribute(head, PKCS11_CKA_PRIME_2))
+		return PKCS11_CKR_GENERAL_ERROR;
+
+	mbedtls_pk_init(&pk);
+	mbedtls_mpi_init(&n);
+	mbedtls_mpi_init(&e);
+	mbedtls_mpi_init(&d);
+	mbedtls_mpi_init(&p);
+	mbedtls_mpi_init(&q);
+
+	mbedtls_rc = mbedtls_pk_parse_key(&pk, data, key_size, NULL, 0);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	rsa = mbedtls_pk_rsa(pk);
+	mbedtls_rc = mbedtls_rsa_export(rsa, &n, &p, &q, &d, &e);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	if (key_bits && mbedtls_mpi_bitlen(&n) != key_bits) {
+		rc = PKCS11_CKR_DATA_LEN_RANGE;
```

Comment:
Agree PKCS11_CKR_DATA_LEN_RANGE is not correct here,
But we cannot use CKR_UNWRAPPING_KEY_SIZE_RANGE
because it looks like indicates the problem with the key that is used to decrypt a wrapped target key
CKR_UNWRAPPING_KEY_SIZE_RANGE: It indicates that although the requested unwrapping operation could in principle be carried out, this Cryptoki library (or the token) is unable to actually do it because the supplied key’s size is outside the range of key sizes that it can handle.

Looks like CKR_WRAPPED_KEY_LEN_RANGE
is a better choice
CKR_WRAPPED_KEY_LEN_RANGE: This value can only be returned by C_UnwrapKey. It
indicates that the provided wrapped key can be seen to be invalid solely on the basis of its length.
This return value has higher priority than CKR_WRAPPED_KEY_INVALID.

https://docs.oasis-open.org/pkcs11/pkcs11-spec/v3.1/cs01/pkcs11-spec-v3.1-cs01.html#_Toc111203249

Reply:
done